[view-transitions] Allow animations to run more than once

nt1m opened the pull request:
#### 4ea01d73bca34b0351de5cf75c58ebb3ce72dd6a
<pre>
[view-transitions] Allow animations to run more than once
<a href="https://bugs.webkit.org/show_bug.cgi?id=269460">https://bugs.webkit.org/show_bug.cgi?id=269460</a>
<a href="https://rdar.apple.com/123005914">rdar://123005914</a>

Reviewed by Antoine Quint.

We need to cancel animations when the view transition ends, so they can be re-ran when the view transition is restarted.

Also add (incomplete) code to sort web animations for view transition pseudo-elements in tree order to avoid assertions.

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-element-animations-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-element-animations.html: Added.
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::compareStyleOriginatedAnimationOwningElementPositionsInDocumentTreeOrder):
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::clearViewTransition):

Canonical link: <a href="https://commits.webkit.org/274734@main">https://commits.webkit.org/274734@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6a6d9a55f56a9ee6df94af1e388e8e5f1c4cb21c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39871 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18882 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42246 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42415 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35773 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42178 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/21808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16212 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33235 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40445 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15916 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34474 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13775 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13799 "9 flakes 1 failures") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35430 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43694 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36205 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35766 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39537 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14686 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12083 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16305 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/34708 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/8945 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16354 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5250 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15949 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10417,6 +10417,7 @@ void Document::performPendingViewTransitions()
 {
     if (!m_activeViewTransition)
         return;
+
```
This should be removed.

graouts commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -377,6 +378,18 @@ void ViewTransition::clearViewTransition()
 
     // FIXME: Implement step 3.
 
+
```
Extra blank line here.

nt1m committed:
[view-transitions] Allow animations to run more than once
https://bugs.webkit.org/show_bug.cgi?id=269460
rdar://123005914

Reviewed by Antoine Quint.

We need to cancel animations when the view transition ends, so they can be re-ran when the view transition is restarted.

Also add (incomplete) code to sort web animations for view transition pseudo-elements in tree order to avoid assertions.

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-element-animations-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-element-animations.html: Added.
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::compareStyleOriginatedAnimationOwningElementPositionsInDocumentTreeOrder):
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::clearViewTransition):

Canonical link: https://commits.webkit.org/274734@main
